### PR TITLE
[PM-3169] Login decryption options in extension popup

### PR DESCRIPTION
**Pull request by coroiu**
## Type of change

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective
[PM-3169](https://bitwarden.atlassian.net/browse/PM-3169)

Allow the clients to differentiate between locked and not-yet-decrypted states and redirect to the corresponding components. This PR does a few things:

- Refactors `lockGuard` and adds a new `redirectGuard` to consolidate redirection logic and modernize guards.
- Moves org identifier to state.
- Adds new state variable to check if we've ever had a user key.
- Modifies routing so that we are able to navigate to decryption options on client startup e.g. when opening an extension, but also works when switching accounts on desktop too (I know cool right, was not my intention :D)

### Popup navigation
- **apps/browser/src/auth/popup/sso.component.ts:**  Close window instead of navigating to decryption options
- **apps/browser/src/auth/popup/two-factor.component.ts:** Close window instead of navigating to decryption options

### Save and load identifier to state
- **libs/angular/src/auth/components/sso.component.ts:** Save identifier to state after successful login. Removed identifier from query params.
- **libs/angular/src/auth/components/two-factor.component.ts:** Removed identifier from query params.
- **libs/angular/src/auth/components/base-login-decryption-options.component.ts:** Get identifier from state instead of query params

### New guards
- **libs/angular/src/auth/guards/lock.guard.ts:** Complete refactor of guard. Converted into functional guard. Redirect logic outsourced to `redirectGuard`. Has special exception to allow navigation to lock page from login-initiated page as part of the TDE Login with MP flow. 
- **libs/angular/src/auth/guards/tde-decryption-required.guard.ts:** Guard that only allow access to the login-initiated page if the vault is locked and has never been decrypted.
- **libs/angular/src/auth/guards/redirect.guard.ts:** New cross-platform guard that contains all redirect logic.
- **apps/web/src/app/guards/home.guard.ts:** Removed and replaced with cross-platform `redirectGuard`
- **libs/angular/src/services/jslib-services.module.ts:** Remove old `LockGuardService`
- **apps/browser/src/popup/services/services.module.ts:** Remove old `LockGuardService`
- - **libs/common/src/auth/services/device-trust-crypto.service:** Add new method to check if current user has TDE enabled. Used in guards.

### Use new guards in routing
- **apps/browser/src/popup/app-routing.module.ts:** Add `tdeDecryptionRequiredGuard()`
- **apps/desktop/src/app/app-routing.module.ts:** Add new root `redirectGuard`. Add `tdeDecryptionRequiredGuard()`
- **apps/web/src/app/oss-routing.module.ts:** Add new root `redirectGuard`. Add `tdeDecryptionRequiredGuard()`

### `everHadUserKey`
- **libs/common/src/platform/abstractions/crypto.service.ts:** Add `everHadUserKey` support.
- **libs/common/src/platform/services/crypto.service.ts:** Add `everHadUserKey` support.

### State stuff
- **libs/common/src/platform/abstractions/state.service.ts:** Add `everHadUserKey` and `userSsoOrganizationIdentifier` support.
- - **libs/common/src/platform/services/state.service.ts:** Add `everHadUserKey` and `userSsoOrganizationIdentifier` support.
- **libs/common/src/platform/models/domain/account.ts:** Add `everHadUserKey` and `userSsoOrganizationIdentifier` support.

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

[PM-3169]: https://bitwarden.atlassian.net/browse/PM-3169?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by JaredSnider-Bitwarden — APPROVED** [review]
Tested all PM-3169 scenarios locally, and all the issues were resolved. Required fixing the TDE Login with MP flow w/ a small edit to the `LockGuard`, but overall, this was an incredibly well done refactor of our guard logic. Thank you for your work on this!

File: libs/angular/src/auth/components/sso.component.ts [hunk 1]
```diff
@@ -73,11 +74,9 @@ export class SsoComponent {
...
           this.checkState(state, qParams.state)
         ) {
-          await this.logIn(
-            qParams.code,
-            codeVerifier,
-            this.getOrgIdentifierFromState(qParams.state)
-          );
+          const ssoOrganizationIdentifier = this.getOrgIdentifierFromState(qParams.state);
+          await this.logIn(qParams.code, codeVerifier, ssoOrganizationIdentifier);
+          await this.stateService.setUserSsoOrganizationIdentifier(ssoOrganizationIdentifier);
```
Comment (justindbaur) [inline libs/angular/src/auth/components/sso.component.ts:79]: Do I have these steps right:
1. Save organization identifier into state
2. Navigate to login decryption options
3. Pull identifier from state
4. Use it to check the enroll status of the user
5. Never delete that information

If that is the case can we pass that information to the next page via query parameter?

File: libs/common/src/platform/abstractions/state.service.ts [hunk 2]
```diff
@@ -475,6 +477,11 @@ export abstract class StateService<T extends Account = Account> {
   setSsoOrganizationIdentifier: (value: string, options?: StorageOptions) => Promise<void>;
   getSsoState: (options?: StorageOptions) => Promise<string>;
   setSsoState: (value: string, options?: StorageOptions) => Promise<void>;
+  getUserSsoOrganizationIdentifier: (options?: StorageOptions) => Promise<string>;
+  setUserSsoOrganizationIdentifier: (
+    value: Partial<string>,
```
Comment (justindbaur) [inline libs/common/src/platform/abstractions/state.service.ts:482]: Why the `Partial<string>`? TypeScript seems to just resolve it to `string`. https://www.typescriptlang.org/play?#code/C4TwDgpgBASlC8UAKBDATsAligNgHgGdg1MA7AcwD4BuIA

**Review by justindbaur — CHANGES_REQUESTED** [review]
A few questions and a few change requests but great work (to both you and Andreas).

File: libs/angular/src/auth/components/base-login-decryption-options.component.ts [hunk 3]
```diff
@@ -107,11 +107,9 @@ export class BaseLoginDecryptionOptionsComponent implements OnInit, OnDestroy {
         // We are dealing with a new account if:
         //  - User does not have admin approval (i.e. has not enrolled into admin reset)
         //  - AND does not have a master password
-        this.platformUtilsService.showToast(
-          "success",
-          null,
-          this.i18nService.t("accountSuccessfullyCreated")
-        );
+
+        // TODO: discuss how this doesn't make any sense to show here
```
Comment (justindbaur) [inline libs/angular/src/auth/components/base-login-decryption-options.component.ts:111]: Just be sure to have that discussion.

**Review by justindbaur — APPROVED** [review]
Thank you for all the changes and our discussion. Looks great.

**Pull request merged**